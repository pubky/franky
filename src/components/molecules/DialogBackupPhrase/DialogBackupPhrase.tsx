'use client';

import { useState, useEffect } from 'react';
import Image from 'next/image';

import * as Atoms from '@/atoms';
import * as Libs from '@/libs';
import * as Stores from '@/core';
import * as Molecules from '@/molecules';

export function DialogBackupPhrase() {
  const [isHidden, setIsHidden] = useState(true);
  const [recoveryWords, setRecoveryWords] = useState<string[]>([]);
  const [step, setStep] = useState(1);
  const { mnemonic } = Stores.useOnboardingStore();

  const handleClose = () => {
    setIsHidden(true);
    setStep(1);
  };

  useEffect(() => {
    if (mnemonic) {
      // Use existing mnemonic (generated by PublicKeyCard)
      setRecoveryWords(mnemonic.split(' '));
    }
  }, [mnemonic]);

  return (
    <Atoms.Dialog>
      <Atoms.DialogTrigger asChild>
        <Atoms.Button id="backup-recovery-phrase-btn" variant="secondary" className="gap-2">
          <Libs.FileText className="h-4 w-4" />
          <span>Recovery phrase</span>
        </Atoms.Button>
      </Atoms.DialogTrigger>
      <Atoms.DialogContent className="gap-6 p-8">
        {step === 1 && (
          <RecoveryStep1
            recoveryWords={recoveryWords}
            isHidden={isHidden}
            setIsHidden={setIsHidden}
            setStep={setStep}
          />
        )}
        {step === 2 && <RecoveryStep2 recoveryWords={recoveryWords} setStep={setStep} />}
        {step === 3 && <RecoveryStep3 handleClose={handleClose} />}
      </Atoms.DialogContent>
    </Atoms.Dialog>
  );
}

function RecoveryStep1({
  recoveryWords,
  isHidden,
  setIsHidden,
  setStep,
}: {
  recoveryWords: string[];
  isHidden: boolean;
  setIsHidden: (isHidden: boolean) => void;
  setStep: (step: number) => void;
}) {
  return (
    <>
      <Atoms.DialogHeader className="space-y-1.5 pr-6">
        <Atoms.DialogTitle className="text-2xl font-bold leading-8 sm:text-xl sm:leading-7">
          Backup recovery phrase
        </Atoms.DialogTitle>
        <Atoms.DialogDescription className="text-sm leading-5 max-w-[530px]">
          Use the recovery phrase below to recover your account at a later date. Write down these 12 words in the
          correct order and store them in a safe place.{' '}
          <span className="text-brand font-bold">Never share this recovery phrase with anyone.</span>
        </Atoms.DialogDescription>
      </Atoms.DialogHeader>

      <Atoms.Container className={Libs.cn(isHidden && 'blur-xs')}>
        <Atoms.Container display="grid" className="grid-cols-2 md:grid-cols-3 gap-3">
          {recoveryWords.map((word, index) => (
            <Atoms.Container key={index} className="items-center gap-3 rounded-md bg-secondary p-4 flex-row">
              <Atoms.Badge
                id={`backup-recovery-phrase-word-${index + 1}`}
                variant="outline"
                className="h-5 min-w-[20px] rounded-full px-1 font-semibold"
              >
                {index + 1}
              </Atoms.Badge>
              <span className="text-base font-medium">{word}</span>
            </Atoms.Container>
          ))}
        </Atoms.Container>
      </Atoms.Container>

      <Atoms.Container className="gap-4 sm:gap-3 flex-row justify-between">
        {isHidden ? (
          <>
            <Atoms.DialogClose asChild>
              <Atoms.Button
                id="backup-recovery-phrase-cancel-btn"
                variant="outline"
                className="h-[60px] flex-1 rounded-full sm:h-10 px-12 py-6"
              >
                Cancel
              </Atoms.Button>
            </Atoms.DialogClose>
            <Atoms.Button
              id="backup-recovery-phrase-reveal-btn"
              className="h-[60px] flex-1 rounded-full sm:h-10 px-12 py-6"
              onClick={() => {
                setIsHidden(!isHidden);
                setStep(1);
              }}
            >
              <Libs.Eye className="mr-2 h-4 w-4" />
              Reveal recovery phrase
            </Atoms.Button>
          </>
        ) : (
          <>
            <Atoms.Button
              variant="outline"
              className="h-[60px] flex-1 rounded-full sm:h-10 px-12 py-6"
              onClick={() => {
                setIsHidden(!isHidden);
                setStep(1);
              }}
            >
              <Libs.EyeOff className="mr-2 h-4 w-4" />
              Hide recovery phrase
            </Atoms.Button>
            <Atoms.Button
              id="backup-recovery-phrase-confirm-btn"
              className="h-[60px] flex-1 rounded-full sm:h-10 px-12 py-6"
              onClick={() => setStep(2)}
            >
              <Libs.ArrowRight className="mr-2 h-4 w-4" />
              Confirm recovery phrase
            </Atoms.Button>
          </>
        )}
      </Atoms.Container>
    </>
  );
}

function RecoveryStep2({ recoveryWords, setStep }: { recoveryWords: string[]; setStep: (step: number) => void }) {
  const [userWords, setUserWords] = useState<string[]>(Array(12).fill(''));
  const [errors, setErrors] = useState<boolean[]>(Array(12).fill(false));
  const [availableWords] = useState<string[]>([...recoveryWords].sort());
  const [isLargeScreen, setIsLargeScreen] = useState(false);

  useEffect(() => {
    const checkScreenSize = () => {
      setIsLargeScreen(window.innerWidth >= 640);
    };

    checkScreenSize();
    window.addEventListener('resize', checkScreenSize);

    return () => window.removeEventListener('resize', checkScreenSize);
  }, []);

  const validateSingleWord = (wordIndex: number, word: string) => {
    const newErrors = [...errors];
    // Check if this specific word is correct for its position
    const isError = word !== '' && word !== recoveryWords[wordIndex];
    newErrors[wordIndex] = isError;
    setErrors(newErrors);
  };

  const handleWordClick = (word: string) => {
    // Check if word is already used
    const isWordAlreadyUsed = userWords.includes(word);

    if (isWordAlreadyUsed) {
      return;
    }

    const emptyIndex = userWords.findIndex((w) => w === '');
    if (emptyIndex !== -1) {
      const newUserWords = [...userWords];
      newUserWords[emptyIndex] = word;
      setUserWords(newUserWords);

      // Validate this specific word immediately
      validateSingleWord(emptyIndex, word);
    }
  };

  const validateWords = () => {
    // Validate all words first
    const newErrors = userWords.map((word, index) => {
      return word !== '' && word !== recoveryWords[index];
    });
    setErrors(newErrors);

    // Check if all words are correct and filled
    const allFilled = userWords.every((word) => word !== '');
    const allCorrect = newErrors.every((error) => !error);

    // Only advance if all words are filled and correct
    if (allFilled && allCorrect) {
      setStep(3);
    }
  };

  const clearWord = (index: number) => {
    const word = userWords[index];
    if (word) {
      const newUserWords = [...userWords];
      newUserWords[index] = '';
      setUserWords(newUserWords);

      // Clear error for this specific word
      const newErrors = [...errors];
      newErrors[index] = false;
      setErrors(newErrors);
    }
  };

  return (
    <>
      <Atoms.DialogHeader className="space-y-1.5 pr-6">
        <Atoms.DialogTitle className="text-2xl sm:text-[24px] font-bold">Confirm recovery phrase</Atoms.DialogTitle>
        <Atoms.DialogDescription className="text-sm text-muted-foreground">
          Click or tap the 12 words in the correct order. Click on filled fields to remove words.
        </Atoms.DialogDescription>
      </Atoms.DialogHeader>

      <Atoms.Container className="space-y-6">
        <Atoms.Container className="flex-wrap gap-2 flex-row">
          {availableWords.map((word, index) => {
            const isUsed = userWords.includes(word);
            return (
              <Atoms.Button
                id={`backup-recovery-phrase-word-${word}`}
                key={`${word}-${index}`}
                variant={isUsed ? 'secondary' : 'outline'}
                size="sm"
                className={`rounded-full ${
                  isUsed
                    ? 'opacity-40 bg-transparent border text-muted-foreground cursor-not-allowed'
                    : 'dark:border-transparent bg-secondary cursor-pointer'
                }`}
                onClick={() => !isUsed && handleWordClick(word)}
                disabled={isUsed}
              >
                {word}
              </Atoms.Button>
            );
          })}
        </Atoms.Container>

        <Atoms.Container display="grid" className="grid-cols-1 sm:grid-cols-3 gap-3">
          {userWords.map((word, i) => {
            const isCorrect = word !== '' && word === recoveryWords[i];
            const isError = errors[i];
            return (
              <Molecules.WordSlot
                key={i}
                mode="readonly"
                index={i}
                word={word}
                isCorrect={isCorrect}
                isError={isError}
                onClear={clearWord}
              />
            );
          })}
        </Atoms.Container>
      </Atoms.Container>

      <Atoms.Container className="flex-col-reverse sm:flex-row gap-3 sm:gap-4 sm:justify-end">
        <Atoms.Button
          variant="outline"
          className="rounded-full flex-1"
          size={isLargeScreen ? 'lg' : 'default'}
          onClick={() => setStep(1)}
        >
          <Libs.ArrowLeft className="mr-2 h-4 w-4" />
          Back
        </Atoms.Button>
        <Atoms.Button
          id="backup-recovery-phrase-validate-btn"
          className="rounded-full flex-1"
          size={isLargeScreen ? 'lg' : 'default'}
          onClick={validateWords}
          disabled={userWords.some((word) => word === '') || errors.some((error) => error)}
        >
          <Libs.Check className="mr-2 h-4 w-4" />
          Validate
        </Atoms.Button>
      </Atoms.Container>
    </>
  );
}

function RecoveryStep3({ handleClose }: { handleClose: () => void }) {
  return (
    <>
      <Atoms.DialogHeader className="pr-6 space-y-1.5">
        <Atoms.DialogTitle className="text-xl lg:text-2xl font-bold leading-7 lg:leading-8">
          Backup complete
        </Atoms.DialogTitle>
        <Atoms.DialogDescription className="text-sm font-medium text-muted-foreground leading-5">
          You can use your backed up recovery phrase to restore your account later.
        </Atoms.DialogDescription>
      </Atoms.DialogHeader>

      <Atoms.Container>
        <Atoms.Container className="w-full bg-card rounded-md p-6 flex items-center justify-center">
          <Image src="/images/check.png" alt="Backup Complete" width={180} height={180} className="w-48 h-48" />
        </Atoms.Container>
      </Atoms.Container>

      <Atoms.Container className="flex-col-reverse sm:flex-row gap-3 lg:gap-4 sm:justify-end">
        <Atoms.DialogClose asChild>
          <Atoms.Button variant="outline" className="rounded-full lg:h-[60px] lg:px-8 flex-1" onClick={handleClose}>
            Cancel
          </Atoms.Button>
        </Atoms.DialogClose>
        <Atoms.DialogClose asChild>
          <Atoms.Button
            id="backup-recovery-phrase-finish-btn"
            className="rounded-full lg:h-[60px] lg:px-8 flex-1"
            onClick={handleClose}
          >
            <Libs.ArrowRight className="mr-2 h-4 w-4" />
            Finish
          </Atoms.Button>
        </Atoms.DialogClose>
      </Atoms.Container>
    </>
  );
}
