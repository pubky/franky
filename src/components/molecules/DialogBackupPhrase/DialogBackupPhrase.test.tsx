import { describe, it, expect, vi } from 'vitest';
import { render, fireEvent, screen } from '@testing-library/react';
import { DialogBackupPhrase } from './DialogBackupPhrase';

// Mock Next.js Image
vi.mock('next/image', () => ({
  __esModule: true,
  default: ({
    src,
    alt,
    width,
    height,
    className,
  }: {
    src: string;
    alt: string;
    width: number;
    height: number;
    className?: string;
  }) => <img data-testid="next-image" src={src} alt={alt} width={width} height={height} className={className} />,
}));

// Mock stores
vi.mock('@/core', () => ({
  useOnboardingStore: () => ({
    secretKey: 'mock-secret-key',
    mnemonic: 'tube tube resource mass door firm genius parrot girl orphan window world',
  }),
}));

// Mock libs partially, preserving real exports and overriding only what's needed
vi.mock('@/libs', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/libs')>();
  return {
    ...actual,
    cn: (...inputs: (string | undefined | null | false)[]) => inputs.filter(Boolean).join(' '),
    FileText: ({ className }: { className?: string }) => (
      <div data-testid="file-text-icon" className={className}>
        FileText
      </div>
    ),
    ArrowLeft: ({ className }: { className?: string }) => (
      <div data-testid="arrow-left-icon" className={className}>
        ArrowLeft
      </div>
    ),
    ArrowRight: ({ className }: { className?: string }) => (
      <div data-testid="arrow-right-icon" className={className}>
        ArrowRight
      </div>
    ),
    Eye: ({ className }: { className?: string }) => (
      <div data-testid="eye-icon" className={className}>
        Eye
      </div>
    ),
    EyeOff: ({ className }: { className?: string }) => (
      <div data-testid="eye-off-icon" className={className}>
        EyeOff
      </div>
    ),
    Copy: ({ className }: { className?: string }) => (
      <div data-testid="copy-icon" className={className}>
        Copy
      </div>
    ),
    Check: ({ className }: { className?: string }) => (
      <div data-testid="check-icon" className={className}>
        Check
      </div>
    ),
    Identity: {
      generateSeedWords: vi.fn(() => [
        'word1',
        'word2',
        'word3',
        'word4',
        'word5',
        'word6',
        'word7',
        'word8',
        'word9',
        'word10',
        'word11',
        'word12',
      ]),
    },
    copyToClipboard: vi.fn(),
  };
});

// Mock atoms
vi.mock('@/atoms', () => ({
  Dialog: ({ children }: { children: React.ReactNode }) => <div data-testid="dialog">{children}</div>,
  DialogTrigger: ({ children, asChild }: { children: React.ReactNode; asChild?: boolean }) => (
    <div data-testid="dialog-trigger" data-as-child={asChild}>
      {children}
    </div>
  ),
  DialogContent: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="dialog-content" className={className}>
      {children}
    </div>
  ),
  DialogHeader: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="dialog-header" className={className}>
      {children}
    </div>
  ),
  DialogTitle: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <h2 data-testid="dialog-title" className={className}>
      {children}
    </h2>
  ),
  DialogDescription: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <p data-testid="dialog-description" className={className}>
      {children}
    </p>
  ),
  DialogClose: ({ children, asChild }: { children: React.ReactNode; asChild?: boolean }) => (
    <div data-testid="dialog-close" data-as-child={asChild}>
      {children}
    </div>
  ),
  Button: ({
    children,
    variant,
    className,
    onClick,
    disabled,
    size,
  }: {
    children: React.ReactNode;
    variant?: string;
    className?: string;
    onClick?: () => void;
    disabled?: boolean;
    size?: string;
  }) => (
    <button
      data-testid={`button-${variant || 'default'}`}
      className={className}
      onClick={onClick}
      disabled={disabled}
      data-size={size}
    >
      {children}
    </button>
  ),
  Container: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="container" className={className}>
      {children}
    </div>
  ),
  Card: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="card" className={className}>
      {children}
    </div>
  ),
  Typography: ({ children, size, className }: { children: React.ReactNode; size?: string; className?: string }) => (
    <p data-testid="typography" data-size={size} className={className}>
      {children}
    </p>
  ),
  Badge: ({ children, variant, className }: { children: React.ReactNode; variant?: string; className?: string }) => (
    <span data-testid="badge" data-variant={variant} className={className}>
      {children}
    </span>
  ),
}));

// Mock molecules
vi.mock('@/molecules', () => ({
  WordSlot: ({
    mode,
    index,
    word,
    isCorrect,
    isError,
    onClear,
  }: {
    mode: string;
    index: number;
    word: string;
    isCorrect: boolean;
    isError: boolean;
    onClear: (index: number) => void;
  }) => (
    <div
      data-testid={`word-slot-${index}`}
      data-mode={mode}
      data-word={word}
      data-is-correct={isCorrect}
      data-is-error={isError}
      onClick={() => onClear(index)}
    >
      {word || `Slot ${index + 1}`}
    </div>
  ),
}));

describe('DialogBackupPhrase - Snapshots', () => {
  it('matches snapshot for default DialogBackupPhrase', () => {
    const { container } = render(<DialogBackupPhrase />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('DialogBackupPhrase - Duplicate Words', () => {
  it('should handle duplicate words correctly in recovery phrase', () => {
    const { container } = render(<DialogBackupPhrase />);

    // The mnemonic contains "tube" twice, so we should see it appear twice in the recovery words grid
    const wordContainers = container.querySelectorAll('[data-testid="container"]');
    const recoveryGrid = Array.from(wordContainers).find((container) =>
      container.querySelector('[data-testid="badge"]'),
    );

    const wordSpans = recoveryGrid?.querySelectorAll('span');
    const tubeWords = Array.from(wordSpans || [])
      .filter((span) => !span.getAttribute('data-testid')) // Filter out badges
      .filter((span) => span.textContent === 'tube');

    // Should have 2 instances of "tube" word
    expect(tubeWords).toHaveLength(2);
  });

  it('should display all 12 words in the recovery phrase grid', () => {
    const { container } = render(<DialogBackupPhrase />);

    // Find the recovery words grid (step 1)
    const wordContainers = container.querySelectorAll('[data-testid="container"]');
    const recoveryGrid = Array.from(wordContainers).find((container) =>
      container.querySelector('[data-testid="badge"]'),
    );

    expect(recoveryGrid).toBeTruthy();

    // Should have 12 word containers
    const badges = recoveryGrid?.querySelectorAll('[data-testid="badge"]');
    expect(badges).toHaveLength(12);
  });

  it('should show correct word order in recovery phrase', () => {
    const { container } = render(<DialogBackupPhrase />);

    // The mnemonic is: "tube tube resource mass door firm genius parrot girl orphan window world"
    const expectedWords = [
      'tube',
      'tube',
      'resource',
      'mass',
      'door',
      'firm',
      'genius',
      'parrot',
      'girl',
      'orphan',
      'window',
      'world',
    ];

    const wordContainers = container.querySelectorAll('[data-testid="container"]');
    const recoveryGrid = Array.from(wordContainers).find((container) =>
      container.querySelector('[data-testid="badge"]'),
    );

    const wordSpans = recoveryGrid?.querySelectorAll('span');
    const actualWords = Array.from(wordSpans || [])
      .filter((span) => !span.getAttribute('data-testid')) // Filter out badges
      .map((span) => span.textContent);

    expect(actualWords).toEqual(expectedWords);
  });

  it('should allow selecting duplicate words individually in step 2', () => {
    const { container } = render(<DialogBackupPhrase />);

    // First, reveal the recovery phrase and go to step 2
    const revealButton = screen.getByText('Reveal recovery phrase');
    fireEvent.click(revealButton);

    const confirmButton = screen.getByText('Confirm recovery phrase');
    fireEvent.click(confirmButton);

    // Now we should be in step 2 with the word selection buttons
    // The sorted words should be: door, firm, genius, girl, mass, orphan, parrot, resource, tube, tube, window, world
    const wordButtons = container.querySelectorAll('button[data-testid^="button-"]');
    const tubeButtons = Array.from(wordButtons).filter((button) => button.textContent?.includes('tube'));

    // Should have 2 instances of "tube" button
    expect(tubeButtons).toHaveLength(2);

    // Both should be clickable initially
    expect(tubeButtons[0]).not.toBeDisabled();
    expect(tubeButtons[1]).not.toBeDisabled();

    // Click the first tube button
    fireEvent.click(tubeButtons[0]);

    // The first button should now be disabled, but the second should still be clickable
    expect(tubeButtons[0]).toBeDisabled();
    expect(tubeButtons[1]).not.toBeDisabled();

    // Click the second tube button
    fireEvent.click(tubeButtons[1]);

    // Now both should be disabled
    expect(tubeButtons[0]).toBeDisabled();
    expect(tubeButtons[1]).toBeDisabled();
  });

  it('should re-enable exact duplicate instance when clearing a slot', () => {
    const { container } = render(<DialogBackupPhrase />);

    // Navigate to step 2
    const revealButton = screen.getByText('Reveal recovery phrase');
    fireEvent.click(revealButton);

    const confirmButton = screen.getByText('Confirm recovery phrase');
    fireEvent.click(confirmButton);

    // Get both tube buttons (duplicates)
    const wordButtons = container.querySelectorAll('button[data-testid^="button-"]');
    const tubeButtons = Array.from(wordButtons).filter((button) => button.textContent?.includes('tube'));

    // Click first tube instance (should go to slot 0)
    fireEvent.click(tubeButtons[0]);
    expect(tubeButtons[0]).toBeDisabled();
    expect(tubeButtons[1]).not.toBeDisabled();

    // Click second tube instance (should go to slot 1)
    fireEvent.click(tubeButtons[1]);
    expect(tubeButtons[0]).toBeDisabled();
    expect(tubeButtons[1]).toBeDisabled();

    // Get the first slot (should contain first tube)
    const wordSlot0 = container.querySelector('[data-testid="word-slot-0"]');
    expect(wordSlot0?.getAttribute('data-word')).toBe('tube');

    // Clear the first slot
    fireEvent.click(wordSlot0!);

    // Only the first tube button should be re-enabled
    expect(tubeButtons[0]).not.toBeDisabled();
    expect(tubeButtons[1]).toBeDisabled();

    // Clear the second slot
    const wordSlot1 = container.querySelector('[data-testid="word-slot-1"]');
    fireEvent.click(wordSlot1!);

    // Now both tube buttons should be enabled again
    expect(tubeButtons[0]).not.toBeDisabled();
    expect(tubeButtons[1]).not.toBeDisabled();
  });

  it('should block validate button when there are errors present', () => {
    const { container } = render(<DialogBackupPhrase />);

    // Navigate to step 2
    const revealButton = screen.getByText('Reveal recovery phrase');
    fireEvent.click(revealButton);

    const confirmButton = screen.getByText('Confirm recovery phrase');
    fireEvent.click(confirmButton);

    // Get all word buttons
    const wordButtons = screen.getAllByRole('button');

    // Click a wrong word for slot 0 (should be "tube", let's click "door")
    const doorBtn = wordButtons.find((btn) => btn.textContent === 'door');
    fireEvent.click(doorBtn!);

    // Verify slot 0 has an error
    const wordSlot0 = container.querySelector('[data-testid="word-slot-0"]');
    expect(wordSlot0?.getAttribute('data-is-error')).toBe('true');

    // Fill remaining slots with any words to enable the button check
    const tubeButtons = wordButtons.filter((button) => button.textContent?.includes('tube'));
    const resourceBtn = wordButtons.find((btn) => btn.textContent === 'resource');
    const massBtn = wordButtons.find((btn) => btn.textContent === 'mass');
    const firmBtn = wordButtons.find((btn) => btn.textContent === 'firm');
    const geniusBtn = wordButtons.find((btn) => btn.textContent === 'genius');
    const parrotBtn = wordButtons.find((btn) => btn.textContent === 'parrot');
    const girlBtn = wordButtons.find((btn) => btn.textContent === 'girl');
    const orphanBtn = wordButtons.find((btn) => btn.textContent === 'orphan');
    const windowBtn = wordButtons.find((btn) => btn.textContent === 'window');
    const worldBtn = wordButtons.find((btn) => btn.textContent === 'world');

    fireEvent.click(tubeButtons[0]);
    fireEvent.click(resourceBtn!);
    fireEvent.click(massBtn!);
    // Skip door since we already used it
    fireEvent.click(firmBtn!);
    fireEvent.click(geniusBtn!);
    fireEvent.click(parrotBtn!);
    fireEvent.click(girlBtn!);
    fireEvent.click(orphanBtn!);
    fireEvent.click(windowBtn!);
    fireEvent.click(worldBtn!);

    // The validate button should be disabled because there's an error
    const validateButton = screen.getByText('Validate').closest('button');
    expect(validateButton).toBeDisabled();
  });

  it('should show errors for incorrectly placed duplicate words', () => {
    const { container } = render(<DialogBackupPhrase />);

    // Navigate to step 2
    const revealButton = screen.getByText('Reveal recovery phrase');
    fireEvent.click(revealButton);

    const confirmButton = screen.getByText('Confirm recovery phrase');
    fireEvent.click(confirmButton);

    // Get tube buttons
    const wordButtons = container.querySelectorAll('button[data-testid^="button-"]');
    const tubeButtons = Array.from(wordButtons).filter((button) => button.textContent?.includes('tube'));

    // Click second tube first (wrong for slot 0)
    fireEvent.click(tubeButtons[1]);

    // Check that slot 0 has an error (because any "tube" is correct for slot 0,
    // this test assumes validation marks it as correct. But if we fill wrong words later, errors will show)
    // Actually, since both are "tube", slot 0 would be correct. Let's test with a different word.

    // Clear and test with a completely wrong word
    const wordSlot0 = container.querySelector('[data-testid="word-slot-0"]');
    fireEvent.click(wordSlot0!); // Clear it

    // Click a wrong word for slot 0 (should be "tube", let's click "door")
    const doorBtn = Array.from(wordButtons).find((btn) => btn.textContent === 'door');
    fireEvent.click(doorBtn!);

    // Check that slot 0 shows error
    const slot0AfterWrong = container.querySelector('[data-testid="word-slot-0"]');
    expect(slot0AfterWrong?.getAttribute('data-is-error')).toBe('true');
  });
});
