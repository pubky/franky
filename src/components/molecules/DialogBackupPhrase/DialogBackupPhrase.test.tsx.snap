// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DialogBackupPhrase - Snapshots > matches snapshot for default DialogBackupPhrase 1`] = `
<div
  data-testid="dialog"
>
  <div
    data-as-child="true"
    data-testid="dialog-trigger"
  >
    <button
      class="gap-2"
      data-testid="button-secondary"
    >
      <div
        class="h-4 w-4"
        data-testid="file-text-icon"
      >
        FileText
      </div>
      <span>
        Recovery phrase
      </span>
    </button>
  </div>
  <div
    class="gap-6 p-8"
    data-testid="dialog-content"
  >
    <div
      class="space-y-1.5 pr-6"
      data-testid="dialog-header"
    >
      <h2
        class="text-2xl font-bold leading-8 sm:text-xl sm:leading-7"
        data-testid="dialog-title"
      >
        Backup recovery phrase
      </h2>
      <p
        class="text-sm leading-5 max-w-[530px]"
        data-testid="dialog-description"
      >
        Use the recovery phrase below to recover your account at a later date. Write down these 12 words in the correct order and store them in a safe place.
         
        <span
          class="text-brand font-bold"
        >
          Never share this recovery phrase with anyone.
        </span>
      </p>
    </div>
    <div
      class="blur-xs"
      data-testid="container"
    >
      <div
        class="grid-cols-2 md:grid-cols-3 gap-3"
        data-testid="container"
      />
    </div>
    <div
      class="gap-4 sm:gap-3 md:flex-row justify-between"
      data-testid="container"
    >
      <div
        data-as-child="true"
        data-testid="dialog-close"
      >
        <button
          class="flex-1 rounded-full h-10 px-4 py-2.5 md:px-12 md:py-6"
          data-testid="button-outline"
        >
          Cancel
        </button>
      </div>
      <button
        class="flex-1 rounded-full h-10 px-4 py-2.5 md:px-12 md:py-6"
        data-testid="button-default"
      >
        <div
          class="mr-2 h-4 w-4"
          data-testid="eye-icon"
        >
          Eye
        </div>
        Reveal recovery phrase
      </button>
    </div>
  </div>
</div>
`;
